### PR TITLE
Harden public registration abuse controls

### PR DESCRIPTION
**Pull request by pauljsnider**
Cleared to flow through automation (owner-approved 2026-07-21). All newly added enforcement is passive/observe-only by default: new App Check and rate-limit modes resolve to `observe` when their policy documents are absent, and no runtime-policy or deployment change ships in this PR. Enabling enforcement later (App Check health, rate-limit telemetry, verified-email rollout) remains a separate, gated step.

## What changed

- Add verified App Check observation/enforcement gates to public registration submission, checkout creation, and checkout cancellation; only Firebase-verified callable context is trusted.
- Add durable hashed subject/network/form rate-limit boundaries with staged `disabled` / `observe` / `enforce` modes. The existing 3-per-10-minute submission limit remains enforced.
- Add stable submission idempotency keys across legacy web and React clients, deterministic registration IDs, payload-conflict detection, and exact replay responses without double-consuming capacity or rate limits.
- Bound public field counts, key/value sizes, aggregate payload sizes, and quantity before Firestore work.
- Attribute authenticated submissions by UID and stage verified-email read enforcement without breaking existing registrations or unverified accounts by default.
- Query parent registrations by both legacy guardian email and authoritative submitter UID, with deduplication and partial-query degradation.
- Add the required collection-group index and a detailed rollout/canary/rollback runbook.

## Trust boundaries and passivity

- Raw App Check headers are ignored; only `context.app.appId` counts as verified.
- Webhook delivery stays exempt from App Check because Stripe authenticates it with signatures and its existing controls.
- New modes resolve to `observe` when missing or invalid. No deployment or runtime-policy change is included here.
- The verified-email Firestore policy remains backward compatible while its policy document is absent or set to observe.
- No data migration is required; existing clients and registration documents remain supported.

## Validation

- `npm test -- --reporter=dot` — 4,566 passed, 81 skipped
- `npm test --prefix apps/app -- --reporter=dot` — 1,390 passed
- `npm test --prefix functions` — 206 passed
- `npm run app:build` — passed; 736-node bundle visualizer verified
- `npm run test:storage-rules:ci` — 71 passed against Firestore/Storage emulators
- `npm run ci:firebase-rules` — passed
- `npx playwright test tests/smoke/registration.spec.js` — 8 passed against a local server
- `git diff --check` and Node syntax checks — passed

## Required rollout order

1. Deploy in observe mode only and collect one normal registration cycle of telemetry.
2. Enforce checkout-subject limits, then network limits, one deploy at a time.
3. Keep form-wide enforcement optional unless measured peak volume proves safe.
4. Enforce App Check only after web/iOS/Android token health is acceptable.
5. Stage verified-email policy in observe, audit, then enforce separately.

Base audited: `17094ca05a5859c43c8ca4b7c967eaa0d6a046d2`
Head for review: `51b2a53ce0c3ed1f4a97c3d31aa887296c42cdbc`

### CODE REVIEW

**Comment by pauljsnider** [comment]
@codex review exact commit df51cfba5342d67a408b5b39831056d439c184cc. Please review the full public-registration abuse-control diff deeply for regressions, passivity/default-observe behavior, idempotency and capacity invariants, App Check trust boundaries, rate-limit evasion/DoS risk, Firestore rule compatibility, and test gaps. This PR must remain draft and must not be merged.

**Comment by pauljsnider** [comment]
@codex review exact commit 458657148a34ad37bda55de48bb3a2976dc75d29. This supersedes the earlier review request after propagating the required legacy cache keys. Please deeply review the full PR for regressions, passive/default-observe behavior, idempotency and capacity invariants, App Check trust boundaries, rate-limit evasion/DoS risk, Firestore rule compatibility, and test gaps. Keep the PR draft and do not merge.

**Comment by pauljsnider** [comment]
@codex review exact commit 3c57109461f100e2e408032a103c5dbe220fe916. This exact head fixes both prior Codex findings: unkeyed legacy clients now call collection.doc() with zero arguments, and checkout subject limits no longer include IP (the separate network boundary still does). Regressions are included and all 203 Functions tests pass. Please re-review the full held draft; do not merge.

File: functions/index.js [hunk 1]
```diff
@@ -883,37 +920,178 @@ async function assertPublicRegistrationRateLimit(input, context = {}) {
...
   } catch (error) {
     throwPublicRegistrationError('invalid-argument', error.message || 'Invalid registration submission.');
   }
 
   const formRef = buildRegistrationFormRef(input);
-  const initialFormSnap = await formRef.get();
-  if (!initialFormSnap.exists) {
-    throwPublicRegistrationError('not-found', 'Registration form not found.');
+  const deterministicRegistrationId = buildPublicRegistrationDocumentId(input);
+  const registrationRef = formRef.collection('registrations').doc(deterministicRegistrationId || undefined);
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/index.js]: Fixed in 3c57109461f100e2e408032a103c5dbe220fe916. The code now branches between registrationsRef.doc(deterministicRegistrationId) and a true zero-argument registrationsRef.doc(). The Firestore test double now rejects an explicitly supplied undefined path, so the existing older/unkeyed-client integration test is a regression for this SDK behavior.

File: functions/public-registration-abuse-core.cjs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    input.publicCheckoutCapability || input.checkoutAttemptToken || input.submissionIdempotencyKey || input.registrationId,
+    'no-subject'
+  );
+  const subject = operation === 'submit' ? guardianEmail : checkoutSubject;
+  const teamId = normalizeBoundaryPart(input.teamId, 'no-team');
+  const formId = normalizeBoundaryPart(input.formId, 'no-form');
+  const appCheckState = getVerifiedAppCheckAppId(context) ? 'verified-app' : 'unverified-app';
+
+  return {
+    subject: ['public-registration', operation, 'subject', teamId, formId, subject, requestIp].join('|'),
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/public-registration-abuse-core.cjs]: Fixed in 3c57109461f100e2e408032a103c5dbe220fe916. Checkout/cancel subject boundaries now omit requestIp, while the independent network boundary remains IP-scoped. A new core test proves the subject is invariant and the network boundary changes when the client rotates IPs.

File: functions/public-registration-abuse-core.cjs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  MAX_PUBLIC_REGISTRATION_FIELD_VALUE_LENGTH,
+  MAX_PUBLIC_REGISTRATION_QUANTITY,
+  assertPublicRegistrationInputLimits,
+  buildPublicRegistrationDocumentId,
+  buildPublicRegistrationRateLimitBoundaries,
+  buildPublicRegistrationSubmissionFingerprint,
+  evaluatePublicRegistrationAppCheck,
+  getVerifiedAppCheckAppId,
+  normalizePublicRegistrationIdempotencyKey,
+  normalizePublicRegistrationSecurityMode
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/public-registration-abuse-core.cjs:180]: No change is applicable: this comment references isSubmissionIdChanged and a check/update sequence that do not exist in this file or PR. Submission idempotency uses a deterministic document ID and rechecks that exact document inside the same Firestore transaction that writes the registration and reserves capacity, covering concurrent requests atomically. Integration tests verify one registration and one capacity reservation.

File: functions/public-registration-abuse-core.cjs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  if (entries.length > MAX_PUBLIC_REGISTRATION_FIELDS) {
+    throw new Error(`${label} details contain too many fields.`);
+  }
+
+  let totalBytes = 0;
+  for (const [key, value] of entries) {
+    if (RESERVED_OBJECT_KEYS.has(key) || !key || key.length > MAX_PUBLIC_REGISTRATION_FIELD_KEY_LENGTH) {
+      throw new Error(`${label} details contain an invalid field.`);
+    }
+    if (typeof value !== 'string' || value.length > MAX_PUBLIC_REGISTRATION_FIELD_VALUE_LENGTH) {
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/public-registration-abuse-core.cjs:121]: No change is applicable: there is no submissionIdBytes, lastSubmissionIdBytes, or Buffer reference comparison in this PR. Idempotency compares canonical SHA-256 fingerprints as lowercase hex strings and rejects a changed payload with already-exists/idempotency-conflict. This thread is attached to field-size validation code.

File: functions/public-registration-abuse-core.cjs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      throw new Error(`${label} details contain an invalid value.`);
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/public-registration-abuse-core.cjs:122]: No change is applicable: this implementation does not store or decode Base64 submission IDs and has no lastSubmissionId field. Keys are validated against a bounded URL-safe pattern, deterministic IDs/fingerprints are SHA-256 hex, and the referenced line is Buffer.byteLength used only to enforce aggregate UTF-8 payload size.

**Comment by pauljsnider** [comment]
@codex review exact commit eff01f8deb82f6d169c6df4b5278c1e2103716b9. This exact head adds only the strict React hook dependencies required by CI after 3c57109461f100e2e408032a103c5dbe220fe916; lint, typecheck, focused registration tests, and all 203 Functions tests pass. Please perform the final exact-head review. Keep this PR draft and do not merge.

**Comment by pauljsnider** [comment]
@codex review exact commit f41d0dc1c04f0dd81904f5c916394fe8a7094055. This exact head addresses both eff01f8deb findings: submit now validates that the form exists and is published before durable limiter reservations, and checkout/cancel prove the registration target before durable limiter reservations; transactions still recheck target state. Missing/unpublished/missing-registration regressions prove zero limiter documents, and all 205 Functions tests pass. Please perform the final exact-head review. Keep this PR draft and do not merge.

File: functions/index.js [hunk 6]
```diff
@@ -883,37 +920,181 @@ async function assertPublicRegistrationRateLimit(input, context = {}) {
...
+      return buildPublicRegistrationReplayResult(
+        registrationRef,
+        existingRegistrationSnap.data() || {},
+        submissionFingerprint
+      );
+    }
   }
 
   await assertPublicRegistrationRateLimit(input, context);
+  await applyStagedPublicRegistrationRateLimits(input, context, 'submit');
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/index.js]: Fixed in f41d0dc1c04f0dd81904f5c916394fe8a7094055. Exact keyed replays are still handled first, then a read-only form lookup verifies existence and published state before the legacy subject limiter or staged durable limiters can write. The transaction retains its existence/published rechecks for close/unpublish races. New missing and unpublished tests assert zero limiter documents.

File: functions/index.js [hunk 7]
```diff
@@ -4173,14 +4356,17 @@ exports.refundStripeTeamFeePayment = functions.https.onCall(async (data, context
...
+exports.createStripeRegistrationCheckout = functions.https.onCall(async (data, context = {}) => {
+  assertPublicRegistrationAppCheck(context, 'create-checkout');
   let input;
   try {
     input = normalizeRegistrationCheckoutInput(data || {});
   } catch (error) {
     throw new functions.https.HttpsError('invalid-argument', error.message || 'Invalid registration checkout request.');
   }
 
+  await applyStagedPublicRegistrationRateLimits(input, context, 'create-checkout');
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/index.js]: Fixed in f41d0dc1c04f0dd81904f5c916394fe8a7094055. Checkout now resolves the capability/path and proves both form and registration exist before staged durable limits; cancel resolves and proves the registration exists before staged durable limits. The existing checkout/cancel transaction logic still revalidates authority/state. A regression exercises both missing targets and asserts zero limiter documents.

**Comment by pauljsnider** [comment]
@codex review exact commit 51b2a53ce0c3ed1f4a97c3d31aa887296c42cdbc. This exact head addresses the f41d0dc1c capability-probe finding with a split limiter: a hashed network-only 120-per-operation/10-minute lookup bucket (controlled by checkout mode, observe by default) runs before resolution, while attacker-keyed subject/network/form buckets remain after target proof. Rotating fake targets reuse exactly one lookup document per IP/operation; enforcement rejects before the 121st lookup. TTL cleanup remains configured. All 206 Functions tests pass. Please perform the final exact-head review. Keep this PR draft and do not merge.

File: functions/index.js [hunk 8]
```diff
@@ -4194,6 +4384,8 @@ exports.createStripeRegistrationCheckout = functions.https.onCall(async (data) =
     throw new functions.https.HttpsError('not-found', 'Registration not found.');
   }
 
+  await applyStagedPublicRegistrationRateLimits(resolvedInput, context, 'create-checkout');
```
[the remark this answers is not in the document]
Comment (pauljsnider) [reply functions/index.js]: Fixed in 51b2a53ce0c3ed1f4a97c3d31aa887296c42cdbc. Checkout and cancel now reserve a pre-resolution lookup-network bucket keyed only by trusted client IP and operation, so rotating capabilities/team/form/registration IDs cannot fan out limiter documents. Checkout mode controls the bucket (observe by default), enforcement rejects before lookup after 120 per ten minutes, and the proven subject limit remains 12. The shared limiter collection already has expiresAt TTL. Regressions prove fake targets reuse one document and the 121st request is rejected before target lookup.

**Comment by pauljsnider** [comment]
@codex review exact commit 51b2a53ce0c3ed1f4a97c3d31aa887296c42cdbc. Retry the final exact-head review because the previous request completed without publishing either findings or a clean reaction. Please review limiter-key cardinality and TTL cleanup, concurrent limiter/idempotency/capacity invariants, passive defaults, and regression coverage. Keep this PR draft and do not merge.

**Comment by pauljsnider** [comment]
Owner-approved to flow through automation (observe-only). Clearing the hold phrase; the pipeline will rebase, run AI review + CI, and merge in observe mode. Enabling enforcement remains a separate gated step.

**Comment by pauljsnider** [comment]
Reviewed current head `da23062e216d`.

paulbot-current-head-review: blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: blocking findings

- **P2 — `functions/index.js:1179`**: Concurrent identical submissions consume the enforced 3-per-10-minute limiter before idempotency is resolved transactionally. Several overlapping retries can exhaust the caller’s quota, causing an exact retry—or a later legitimate submission—to return `resource-exhausted` despite only one registration being created. The added test covers sequential replay only; add a focused concurrent-retry test and resolve/recheck idempotency before rejecting on the limiter.

**Comment by pauljsnider** [comment]
Reviewed current head `a129a32c9a51`.

paulbot-current-head-review: blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: blocking findings

- **P2 — `functions/index.js:1179`**: Concurrent requests using the same idempotency key reserve durable rate-limit slots before the registration transaction detects the replay. Parallel retries therefore increment counters multiple times—and enough retries can return `resource-exhausted` instead of the promised replay response. Move replay detection and rate reservation into an atomic flow or otherwise avoid charging losing concurrent retries; add a focused concurrency test.

**Comment by pauljsnider** [comment]
Reviewed current head `c611e78fc8f1`.

paulbot-current-head-review: blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: blocking findings

- **P2 — `tests/unit/admin-invite-signup-cache-busting.test.js`**: Out-of-scope assertions expect cache versions (`db.js?v=118/119`, `auth.js?v=130`, `utils.js?v=19`) that SHA `c611e78f` does not contain, causing deterministic CI failures.
- **P2 — `tests/unit/team-media-wiring.test.js`**: The test inconsistently expects `db.js?v=118` and `v=119`, while the runtime imports `v=117`; this also deterministically fails CI.

**Comment by pauljsnider** [comment]
Reviewed current head `526e98b96a3a`.

paulbot-current-head-review: blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: blocking findings

- **P2 — `functions/index.js:1155-1193`**: Concurrent identical submissions can double-consume rate-limit capacity. Both requests may miss the preflight registration read, reserve legacy and staged limiter slots, then one becomes an idempotent replay inside the transaction. The focused test at `functions/test/public-registration-submission.test.cjs:581` covers only sequential replay. This violates the PR’s replay-without-consuming-rate-limits guarantee and can prematurely trigger the existing 3-per-10-minute limit during overlapping client/network retries.

**Comment by pauljsnider** [comment]
Reviewed current head `38511fcd5c15`.

paulbot-current-head-review: blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: blocking findings

- **P1 — `functions/index.js:1153`**: Rate-limit reservations use `submissionFingerprint`, which excludes `submissionIdempotencyKey`. An unauthenticated caller can repeatedly send the same payload with fresh idempotency keys and no `checkoutAttemptToken`; each request creates a distinct deterministic registration while all requests reuse one limiter reservation. This bypasses the existing enforced 3-per-10-minute limit and staged limits, enabling registration/capacity exhaustion. Add a focused test proving identical payloads with different keys consume separate slots.

**Comment by pauljsnider** [comment]
Reviewed current head `390ce63b0d67`.

paulbot-current-head-review: no blocking findings

│
◇  Doctor warnings ───────────────────────────────────────────────────────────╮
│                                                                             │
│  - Left 3 Codex binding sidecar(s) in place because session ownership       │
│    is indeterminate: session index                                          │
│    /home/paul-bot1/.openclaw-codex-core/agents/main/sessions/sessions.json  │
│    has invalid entries                                                      │
│                                                                             │
├─────────────────────────────────────────────────────────────────────────────╯
paulbot-current-head-review: no blocking findings

Reviewed current head `390ce63b` against `origin/master`, covering App Check trust, staged rate limits, idempotent replay/concurrency, checkout authority, Firestore rules/indexes, payload bounds, parent readback, rollout defaults, focused tests, and scope. No P0–P2 regression or unsafe boundary change found.
